adrs/0003-testing-strategy.md

```diff
@@ -0,0 +1,49 @@
+# Stacks Testing Strategy
+
+**Date:** 2022-11-28
+
+**Related PR:** [#1195](https://github.com/StackExchange/Stacks/pull/1195)
+
+## Issue
+We currently don't test our Stacks components except from manually/ad-hoc executing some visual regression tests of entire documentation pages with BackstopJS. Linting and formatting despite being setup are also often ignored because they are not running methodically in a pipeline.
+
+**A lack of testing gives us very little confidence refactoring and evolving the codebase.**
+
+## Decision
+
+Introduce 2 initial categories of tests for Stacks to increase confidence working on new and existing components:
+- Component/Unit Tests
+- Component Visual Regression Tests
+
+General guidelines:
+- Tests run in a headless browser context ([wtr + playwright](https://modern-web.dev/docs/test-runner/overview/)) to provide higher level of confidence
+- DOM Querying and Events Simulation are done as much as possible with [testing-library](https://testing-library.com/) and [user-event](https://testing-library.com/docs/user-event/intro/) instead of reaching out directly to DOM APIs
+- Every component should be accompanied by related component and visual regression tests (atomic classes are excluded)
+
+### Component/Unit Tests
+These tests should provide full coverage of the component being tested. Happy path, edge and error cases should be all covered.
+
+Guidelines:
+- Tests should not reach for internal implementation/methods but rather tests the component through its public APIs (from the point of view of a consumer)
+- Complex DOM-agnostic business logic can be extracted in separate functions which can be unit tested in isolation
+
+### Component Visual Regression Tests
+These tests should check how the component will look after any code changes have been made by comparing screenshots taken before and after code changes. 
+
+Guidelines:
+- Baseline images for these tests should be committed to the repository with `git lfs` to avoid bloating the history.
+- When relevant, screenshots at different viewport sizes should be taken
+
+### Additional Guidelines
+
+- All tests and linters runs in a GH workflow for every commit on any branch.
+
+- BackstopJS tests will be gradually retired in favor of the more targeted component visual regression tests approach.
+
+- Tests will be gradually added to existing components as developers need to change them. New components should always be accompanied by tests.
+
+- Consider this is a starting point for Stacks Testing Strategy. More categories, such as a11y testing, might be introduced in future ADRs.
```

Comment:
FWIW, we already test for accessibility issues with mabl, though it's kinda cumbersome to run. I anticipate that [Playwright will lend it self well to a11y testing](https://playwright.dev/docs/accessibility-testing).

Reply:
Agree, we could do all sort of things really: 
- https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot
- https://open-wc.org/docs/testing/chai-a11y-axe/

I would recommend to have a dedicated ADR about accessibility testing though to don't bloat this one up. I will document the task as "operational maintenance" in our internal backlog so that we don't forget.